controllers/cruisecontroltask_controller.go

```diff
@@ -70,121 +82,223 @@ func (r *CruiseControlTaskReconciler) Reconcile(ctx context.Context, request ctr
 		return reconciled()
 	}
 
-	scaler, err := scale.NewCruiseControlScaler(ctx, scale.CruiseControlURLFromKafkaCluster(instance))
+	ccOperationList := banzaiv1alpha1.CruiseControlOperationList{}
+	matchingLabels := client.MatchingLabels(apiutil.LabelsForKafka(instance.Name))
+	err = r.DirectClient.List(ctx, &ccOperationList, client.ListOption(client.InNamespace(request.Namespace)), client.ListOption(matchingLabels))
 	if err != nil {
-		return requeueWithError(log, "failed to create Cruise Control Scaler instance", err)
+		return requeueWithError(log, err.Error(), err)
 	}
-
-	if !scaler.IsUp() {
-		log.Info("requeue event as Cruise Control is not available (yet)")
-		return requeueAfter(DefaultRequeueAfterTimeInSec)
+	// Selecting owned CruiseControlOperations
+	var ccOperations []*banzaiv1alpha1.CruiseControlOperation
+	for i := range ccOperationList.Items {
+		operation := &ccOperationList.Items[i]
+		for _, ownerRef := range operation.GetOwnerReferences() {
+			if ownerRef.Name == instance.Name {
+				ccOperations = append(ccOperations, operation)
+				break
+			}
+		}
 	}
 
 	// Update task states with information from Cruise Control
-	err = updateActiveTasks(scaler, tasksAndStates)
+	err = updateActiveTasks(tasksAndStates, ccOperations)
 	if err != nil {
 		log.Error(err, "requeue event as updating state of active tasks failed")
 		return requeueAfter(DefaultRequeueAfterTimeInSec)
 	}
 
-	// Check if CruiseControl is ready as we cannot perform any operation until it is in ready state
-	if status, _ := scaler.Status(); status.InExecution() {
-		log.Info("updating status of Kafka Cluster and requeue event as Cruise Control is in execution")
-		if err := r.UpdateStatus(ctx, instance, tasksAndStates); err != nil {
-			log.Error(err, "failed to update Kafka Cluster status")
+	if _, ok := r.Scaler.(*scale.MockCruiseControlScaler); !ok {
```

Comment:
For the first question:
Sometimes when the cached client is used the requested resource does not contain the modifications in the status  because of the fast requeue. This can lead duplicated task execution :( I found It safer to use here not caching directClient here.

Reply:
Please also create a refactor ticket for the abstract factory thing.